src/Rules/Compile.hs

```diff
@@ -39,3 +47,35 @@ needCompileDependencies :: Context -> Action ()
 needCompileDependencies context@Context {..} = do
     when (isLibrary package) $ need =<< return <$> pkgConfFile context
     needContext =<< contextDependencies context
+
+needGenerated :: Context -> FilePath -> Action ()
+needGenerated context origFile = go Set.empty
+  where
+    go :: Set.Set String -> Action ()
+    go done = withTempFile $ \outFile -> do
+        let mode = Cc FindMissingInclude $ stage context
+            target = Target context mode [origFile] [outFile]
+        build target
+        deps <- parseFile outFile
+
+        -- Get the full path if the include refers to a generated file and call
+        -- `need` on it.
+        needed <- liftM catMaybes $
+            interpretInContext context (mapM getPathIfGenerated deps)
+        need needed
+
+        let newdone = foldl' (flip Set.insert) done needed
```

Comment:
I tend to find it to be clearer to use: Set.fromList needed `Set.union` done - it's probably more efficient too, but that's very much a side benefit.